### PR TITLE
Add data generator preferences

### PR DESCRIPTION
**Pull request by jcc37**
- UI changes: added data generator control panel to the sidebar
- Added data preference to `PreferenceService`
- Modified `DataService` to provide data subject and reflect data preference

### CODE REVIEW

File: src/services/preference/types.ts [hunk 1]
```diff
@@ -10,6 +10,15 @@ export interface AudificationPreference extends Preference {
   readAfter: boolean;
 }
 
+export interface DataPreference extends Preference {
```
Comment (64json) [inline src/services/preference/types.ts:13]: I think you can make it `export type DataPreference = ModelSettings & Preference`?

File: src/services/data/data.service.ts [hunk 2]
```diff
@@ -7,39 +7,72 @@ import {
...
+    DataService.categories,
+    DataService.measures,
+    DataService.settings,
+  ));
+  private destroy$ = new Subject();
+
+  constructor(private preferenceService: PreferenceService) {
+    this.preferenceService.data$
+      .pipe(takeUntil(this.destroy$))
+      .pipe(throttleTime(1000))
```
Comment (64json) [inline src/services/data/data.service.ts:40]: I think `throttleTime` pipe can be modified a bit to remove kinda redundant initialization of `dataCube$` in line 30.
https://stackoverflow.com/questions/37146081/rxjs-throttle-behavior-get-first-value-immediately

In that case, we can also remove `this.settings`.

File: src/services/preference/types.ts (hunk 1 above)
Comment (jcc37) [reply src/services/preference/types.ts:13]: Actually `DataPreference` is not `DataCube`'s setting. It contains parameters you want users to try to create different datasets. I use a subset of `ModelSettings` just to demonstrate the playground works properly, and more meaningful preference items will be added in future PRs.

File: src/services/data/data.service.ts (hunk 2 above)
Comment (jcc37) [reply src/services/data/data.service.ts:36]: I replaced `BehaviorSubject` with `ReplaySubject` to remove redundant initialization, thanks!

Comment (64json) [reply src/services/data/data.service.ts:36]: I just realized it should be `deboundTime`, not `throttleTime`.

Comment (jcc37) [reply src/services/data/data.service.ts]: `throttleTime` with settings `{ leading: true, trailing: true}` works as expected

File: src/services/data/data.service.ts [hunk 3]
```diff
@@ -50,29 +50,43 @@ export class DataService implements OnDestroy {
...
     this.destroy$.complete();
   }
+
+  observeData(queryOptions: DataQueryOptions) {
+    const {
+      categoryNames$ = new BehaviorSubject([]),
+      measureNames$ = new BehaviorSubject([]),
+      filters$ = new BehaviorSubject(undefined),
+      sortBy$ = new BehaviorSubject(undefined),
+    } = queryOptions;
```
[the remark this answers is not in the document]
Comment (64json) [reply src/services/data/data.service.ts:59]: why haven't I thought that way?!!

File: src/i18n/types.ts [hunk 4]
```diff
@@ -15,8 +15,18 @@ export enum AUDIFICATION {
...
+  avgHits,
+  hitStdDev,
+  avgUsers,
+  userStdDev,
+  avgSessionsPerUser,
+  sessionsPerUserStdDev,
+}
+
+export enum AUDIFICATION_PREFERENCE {
+  enabled = 0x400,
```
Comment (Rhombus33) [inline src/i18n/types.ts:29]: just a question, how is this value used?

File: src/models/data-cube/filters.ts [hunk 5]
```diff
@@ -1,35 +1,27 @@
...
-  const { excludeStartDate, excludeEndDate } = {
-    ...defaultOptions,
-    ...options,
-  };
+export function inOneOfDateRanges(dateRanges: [Date, Date][], rangeOptions: RangeOptions = {}): Filter {
+  const {
+    excludeStart = false,
+    excludeEnd = false,
+  } = rangeOptions;
+  const startOffset = excludeStart ? 1 : 0;
```
Comment (Rhombus33) [inline src/models/data-cube/filters.ts:13]: Could we use a constant for the offset value "1"? To make it clear what time increment it is.

File: src/services/data/data.service.ts [hunk 6]
```diff
@@ -7,40 +6,75 @@ import {
...
-  getMeasureOverDays(measureName: string, days = 30): Datum[] {
-    const categoryName = 'nthDay';
-    const endDate = DateTime.local();
-    const startDate = endDate.minus({ day: days });
+  ngOnDestroy() {
+    this.destroy$.next();
+    this.destroy$.complete();
+  }
+
+  observeData(queryOptions$: Observable<QueryOptions>) {
```
Comment (Rhombus33) [inline src/services/data/data.service.ts:54]: Should be private if that used external to this service.

File: src/models/data-cube/generation.ts [hunk 7]
```diff
@@ -91,8 +93,9 @@ const defaultSettings: ModelSettings = {
   userStdDev: 100,
   avgSessionsPerUser: 5,
   sessionsPerUserStdDev: 3,
-  nthDay: true,
-  days: 60,
+  timeSeries: true,
+  startDate: new Date(Date.now() - 60 * DAY),
```
Comment (Rhombus33) [inline src/models/data-cube/generation.ts:97]: Have constant for 60

File: src/services/data/data.service.ts [hunk 8]
```diff
@@ -7,40 +6,75 @@ import {
...
-        [categoryName],
-        [measureName],
-        [betweenDates(startDate.toJSDate(), endDate.toJSDate(), { excludeStartDate: true })],
-      )
-      .map(row => ({
-        date: startDate
-          .plus({ days: row.categories.get(categoryName) as number })
-          .toJSDate(),
-        value: row.values.get(measureName)!,
+  observeTimeSeries(timeSeriesQueryOptions$: Observable<TimeSeriesQueryOptions>) {
```
Comment (Rhombus33) [inline src/services/data/data.service.ts:61]: Add documentation that this ultimately returns data? or update method name to reflect that?

File: src/components/line-chart/line-chart.component.ts [hunk 9]
```diff
@@ -15,16 +18,20 @@ import { DataService } from '../../services/data/data.service';
 export class LineChartComponent implements RenderOptions, OnChanges, OnInit, OnDestroy {
   @ViewChild(A11yPlaceholderDirective, { static: true }) a11yPlaceholder: A11yPlaceholderDirective<LineChartComponent>;
 
-  @Input() measureName: string;
+  @Input() endDate = new Date();
+  @Input() startDate = new Date(this.endDate.getTime() - 30 * DAY);
+  @Input() measureNames: string[] = [];
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/components/line-chart/line-chart.component.ts:23]: Yes, they are metric names. "Measure" and "Category" are terms used in `DataCube`, and the input here takes the measure names of data provided by `DataService`.

File: src/models/data-cube/data-cube.model.ts [hunk 10]
```diff
@@ -113,17 +113,18 @@ export class DataCube {
...
    * the second, and so on.
    *
-   * @param categoryNames The categories to request a breakdown from.
-   * @param measureNames The measures to provide values for.
-   * @param filters The filters to apply to the cube before finding the results.
-   * @param sortBy The concept names to sort in ascending order.
+   * @param queryOptions.categoryNames The categories to request a breakdown from.
+   * @param queryOptions.measureNames The measures to provide values for.
+   * @param queryOptions.filters The filters to apply to the cube before finding the results.
+   * @param queryOptions.sortBy The concept names to sort in ascending order.
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/models/data-cube/data-cube.model.ts:119]: @64json I think we can move them to where the interface is defined.

File: src/components/line-chart-audification/line-chart-audification.component.ts [hunk 11]
```diff
@@ -165,7 +165,8 @@ export class LineChartAudificationComponent implements AudificationPreference, O
   }
 
   private readOutMeasure() {
-    return this.screenReaderComponent.readOut(humanizeMeasureName(this.measureName));
+    // TODO: read the entire measureNames
+    return this.screenReaderComponent.readOut(humanizeMeasureName(this.measureNames[0]));
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/components/line-chart-audification/line-chart-audification.component.ts:169]: No. `measureNames` here is hard-coded in `dashboard/dashboard.component.html`

File: src/models/data-cube/data-cube.model.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (64json) [reply src/models/data-cube/data-cube.model.ts:119]: Yeah that sounds better

File: src/i18n/languages/en.ts [hunk 12]
```diff
@@ -25,6 +26,14 @@ export const en: I18n = {
   [AUDIFICATION.RANGE]: 'Range from %(min)s to %(max)s',
   [AUDIFICATION.ACTIVE_DATUM]: '%(y)s on %(x)s',
 
+  [DATA_PREFERENCE.enabled]: 'Data Generator',
+  [DATA_PREFERENCE.avgHits]: 'Average Hits',
+  [DATA_PREFERENCE.hitStdDev]: 'Hit Std',
```
[the remark this answers is not in the document]
Comment (jcc37) [reply src/i18n/languages/en.ts:31]: @64json We can modify it to 'Hit Standard Deviation`

File: src/i18n/types.ts (hunk 4 above)
Comment (jcc37) [reply src/i18n/types.ts:29]: `I18nKey` is a union type of all these enums. These values are used to resolve the conflict when defining i18n text for these enums in the same object.

File: src/services/data/data.service.ts (hunk 8 above)
Comment (64json) [reply src/services/data/data.service.ts:61]: This method has been completely moved and renamed in #46 and its child branches. I'll address that in those!

File: src/services/data/data.service.ts (hunk 6 above)
Comment (jcc37) [reply src/services/data/data.service.ts:54]: This part (data retrieval and `DataService`) is thoroughly modified and cleaner in #46 .

File: src/i18n/languages/en.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jcc37) [reply src/i18n/languages/en.ts:31]: @Rhombus33 If I change all "Std" to "Standard Deviation", the text for configurable items in the sidebar would wrap, which makes it ugly. I am thinking to add aria-label for screen readers and keep short for display, but not sure how people deal with this problem. What do you think?

[the remark this answers is not in the document]
Comment (Rhombus33) [reply src/i18n/languages/en.ts:31]: Having a separate aria label SGTM

File: src/i18n/types.ts (hunk 4 above)
Comment (Rhombus33) [reply src/i18n/types.ts:29]: Could you just add a comment describing what these values are for? Thanks!

**Pull request merged**